### PR TITLE
Add token readable kv state

### LINKED ISSUE #9252: Implement `TokenDatabaseReadableKVState`
### Problem

We should add implementation for `TokenDatabaseReadableKVState`, so that we can return tokens in PBJ format from DB.

### Solution

Add `TokenDatabaseReadableKVState` in the following manner:

- Copy `TokenDatabaseAccessor` logic as a starting point
- Add repository logic for reading tokens from DB
- Add a new converter between entity and token DTO types to PBJ, so that the returned object is compliant with the hedera.app state

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by kselveliev**
**Description**:

Adding `TokenReadableKVState` for returning fungible token models into PBJ types, reading from DB.

Added `TokenReadableKVStateTest` which test method in `TokenReadableKVState`

Copied `Token` model from services.
(We need this copy since for some fields we change with Suppliers and we need to modify the hashcode and equals in the copied classes to handle the new supplier changes)

Added `TokenTest` which adds test for all of the Token model methods.

Added couple of utility methods in `CommonEntityAccessor` and `EntityIdUtils`.

Added a new parent class `AbstractStateTest` which has a full token setup logic that is copied from services 

**Related issue(s)**:

Fixes #9252

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Comment by kselveliev** [comment]
Need to add test for the new Token record

File: hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java [hunk 1]
```diff
@@ -0,0 +1,1794 @@
...
+        if (deleted != DEFAULT.deleted) {
+            result = 31 * result + Boolean.hashCode(deleted);
+        }
+        if (tokenType != null && !tokenType.equals(DEFAULT.tokenType)) {
+            result = 31 * result + Integer.hashCode(tokenType.protoOrdinal());
+        }
+        if (supplyType != null && !supplyType.equals(DEFAULT.supplyType)) {
+            result = 31 * result + Integer.hashCode(supplyType.protoOrdinal());
+        }
+        if (autoRenewAccountIdSupplier != null && autoRenewAccountIdSupplier.get() != null) {
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java:311]: What if `autoRenewAccountIdSupplier==null` but `DEFAULT.autoRenewAccountIdSupplier` is not? We won't get inside this if clause.

File: hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java [hunk 2]
```diff
@@ -0,0 +1,1794 @@
...
+        if (totalSupplySupplier != null && !totalSupplySupplier.get().equals(thatObj.totalSupplySupplier.get())) {
+            return false;
+        }
+        if (treasuryAccountIdSupplier == null && thatObj.treasuryAccountIdSupplier != null) {
+            return false;
+        }
+        if (treasuryAccountIdSupplier != null && thatObj.treasuryAccountIdSupplier == null) {
+            return false;
+        }
+        if (treasuryAccountIdSupplier != null) {
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java:414]: What about treasuryAccountIdSupplier == null, but thatObj.treasuryAccountIdSupplier != null? We will be missing this case

File: hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java [hunk 3]
```diff
@@ -0,0 +1,1794 @@
...
+            AccountID thisAccountId = treasuryAccountIdSupplier.get();
+            AccountID thatAccountId = thatObj.treasuryAccountIdSupplier.get();
+
+            // If both retrieved account IDs are null, continue without returning true
+            if (thisAccountId == null && thatAccountId == null) {
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java:419]: We can check if `thisAccount == null && thatAccountId != null` and return false and then in the `else` statement, directly compare  `!thisAccountId.equals(thatAccountId)`, thus having only 2 checks.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/CommonEntityAccessor.java [hunk 4]
```diff
@@ -40,6 +42,22 @@ public class CommonEntityAccessor {
         }
     }
 
+    public @Nonnull Optional<Entity> get(@Nonnull TokenID tokenID, final Optional<Long> timestamp) {
+        return getEntityByMirrorAddressAndTimestamp(
+                EntityId.of(tokenID.shardNum(), tokenID.realmNum(), tokenID.tokenNum()), timestamp);
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/CommonEntityAccessor.java:47]: Use EntityIdUtils for EntityId conversion

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    public long size() {
+        return 0;
+    }
+
+    private Token tokenFromEntities(
+            final Entity entity,
+            final com.hedera.mirror.common.domain.token.Token token,
+            final Optional<Long> timestamp) {
+        return Token.newBuilder()
+                .tokenId(new TokenID(entity.getShard(), entity.getRealm(), entity.getNum()))
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:109]: Sort all builder methods alphabetically.

Comment (IvanKavaldzhiev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:109]: Use EntityIdUtils for TokenID creation

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+    private Supplier<Long> getTotalSupply(
+            final com.hedera.mirror.common.domain.token.Token token, final Optional<Long> timestamp) {
+        return Suppliers.memoize(() -> timestamp
+                .map(t -> getTotalSupplyHistorical(token.getTokenId(), t))
+                .or(() -> Optional.ofNullable(token.getTotalSupply()))
+                .orElse(0L));
+    }
+
+    private Long getTotalSupplyHistorical(long tokenId, long timestamp) {
+        return tokenRepository.findFungibleTotalSupplyByTokenIdAndTimestamp(tokenId, timestamp);
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:148]: This statement should be wrapped in a Suppliers.memoize()

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+        }
+        return Suppliers.memoize(() -> timestamp
+                .map(t -> entityRepository.findActiveByIdAndTimestamp(treasuryId.getId(), t))
+                .orElseGet(() -> entityRepository.findByIdAndDeletedIsFalse(treasuryId.getId()))
+                .map(EntityIdUtils::toAccountId)
+                .orElse(null));
+    }
+
+    private Supplier<List<CustomFee>> getCustomFees(Long tokenId, final Optional<Long> timestamp) {
+        final var customFees = timestamp
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:174]: This assignment should be also wrapped in the Suppliers.memoize()

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+            final var denominatingTokenId = f.getDenominatingTokenId();
+
+            final var fixedFee = new FixedFee(
+                    f.getAmount(),
+                    new TokenID(
+                            denominatingTokenId.getShard(),
+                            denominatingTokenId.getRealm(),
+                            denominatingTokenId.getNum()));
+
+            var constructed = new CustomFee(
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:210]: Use Builder for construction

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+            return Collections.emptyList();
+        }
+
+        var fixedFees = new ArrayList<CustomFee>();
+        customFee.getFixedFees().forEach(f -> {
+            final var collector =
+                    commonEntityAccessor.getAccountWithCanonicalAddress(f.getCollectorAccountId(), timestamp);
+            final var denominatingTokenId = f.getDenominatingTokenId();
+
+            final var fixedFee = new FixedFee(
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:203]: Use builder for construction and EntityIdUtils for TokenID generation

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+            final var collector =
+                    commonEntityAccessor.getAccountWithCanonicalAddress(f.getCollectorAccountId(), timestamp);
+            final var fallbackFee = f.getFallbackFee();
+
+            FixedFee convertedFallbackFee = null;
+            if (fallbackFee != null) {
+                final var denominatingTokenId = fallbackFee.getDenominatingTokenId();
+                convertedFallbackFee = new FixedFee(
+                        fallbackFee.getAmount(),
+                        new TokenID(
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:258]: Use EntityIdUtils for TokenID generation

File: hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java (hunk 3 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java:419]: If we apply the following we are going to get 
**"thisAccountId" is null error** 
in the case where both suppliers return null accountIDs. Or I might be missing something but my test with 2 null values starts to fail. -> `testEqualsWithNullTreasuryAccountIdBoth`

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java (hunk 5 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java]: Comments addressed

File: hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java (hunk 2 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java:414]: line 408 above this does -> 
```
 if (treasuryAccountIdSupplier == null && thatObj.treasuryAccountIdSupplier != null) {
            return false;
 }
```

this should be handling it, right?

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java (hunk 10 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java]: Comment addressed

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java (hunk 9 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java]: Comment addressed

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java (hunk 6 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:149]: Since this method is only called from `getTotalSupply` which is already wrapped in suppliers memoize we might not need it right?

Comment (IvanKavaldzhiev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:149]: You're right

File: hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java (hunk 3 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java:419]: This was my initial implementation but we start getting into null pointers 
since the supplier might not be null but the value it returns can be null
**treasuryAccountIdSupplier.get().equals**

File: hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java [hunk 11]
```diff
@@ -0,0 +1,1794 @@
...
+        if (totalSupplySupplier == null && thatObj.totalSupplySupplier != null) {
+            return false;
+        }
+        if (totalSupplySupplier != null && thatObj.totalSupplySupplier == null) {
+            return false;
+        }
+        if (totalSupplySupplier != null && !totalSupplySupplier.get().equals(thatObj.totalSupplySupplier.get())) {
+            return false;
+        }
+        if (treasuryAccountIdSupplier == null && thatObj.treasuryAccountIdSupplier != null) {
```
Comment (bilyana-gospodinova) [inline hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java:408]: It seems that we have a lot of similar lines - check if one of the suppliers is null and the other is not, then the reverse check, then check if their values differ. We can create a helper utility method that does that (comparing suppliers). This way we will make it easier to read the the `equals()` code and we will prevent accidentally forgetting to add one of the checks.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import org.springframework.util.CollectionUtils;
+
+public class TokenReadableKVState extends ReadableKVStateBase<TokenID, Token> {
+    private static final String KEY = "TOKENS";
```
Comment (bilyana-gospodinova) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:56]: In the other PRs we don't define such constants. Reference: https://github.com/hashgraph/hedera-mirror-node/pull/9450#discussion_r1783574148

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/state/TokenReadableKVStateTest.java [hunk 13]
```diff
@@ -0,0 +1,797 @@
...
+        databaseToken.setTreasuryAccountId(treasuryId);
+        if (timestamp.isPresent()) {
+            when(tokenRepository.findByTokenIdAndTimestamp(entity.getId(), timestamp.get()))
+                    .thenReturn(Optional.ofNullable(databaseToken));
+        } else {
+            when(tokenRepository.findById(any())).thenReturn(Optional.ofNullable(databaseToken));
+        }
+    }
+
+    private AccountID getAccountId(final Long shard, final Long realm, final Long num) {
```
Comment (bilyana-gospodinova) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/state/TokenReadableKVStateTest.java:794]: This can be moved to `EntityIdUtils`.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java (hunk 12 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java]: Yeah i totally forgot, comment addressed.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/state/TokenReadableKVStateTest.java (hunk 13 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/state/TokenReadableKVStateTest.java]: Comment addressed

File: hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java (hunk 11 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/hapi/node/state/token/Token.java]: Pretty good idea, I added a new method in Suppliers util class to compare suppliers.

File: hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityIdUtils.java [hunk 14]
```diff
@@ -180,6 +186,10 @@ public static com.hedera.hapi.node.base.AccountID toAccountId(final EntityId ent
                 .build();
     }
 
+    public static com.hedera.hapi.node.base.AccountID getAccountId(final Long shard, final Long realm, final Long num) {
```
Comment (bilyana-gospodinova) [inline hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityIdUtils.java:189]: The other methods in this class are named `to<class_type>ID`. It will be more readable if we stick to this convention.

Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityIdUtils.java]: Comment addressed

**Review by bilyana-gospodinova — DISMISSED** [review]
LGTM

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+    private final CommonEntityAccessor commonEntityAccessor;
+    private final CustomFeeRepository customFeeRepository;
+    private final TokenRepository tokenRepository;
+    private final EntityRepository entityRepository;
+
+    protected TokenReadableKVState(
+            @Nonnull CommonEntityAccessor commonEntityAccessor,
+            @Nonnull CustomFeeRepository customFeeRepository,
+            @Nonnull TokenRepository tokenRepository,
+            @Nonnull EntityRepository entityRepository) {
```
[the remark this answers is not in the document]
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java]: Comment addressed

File: hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityIdUtils.java [hunk 16]
```diff
@@ -180,6 +186,10 @@ public static com.hedera.hapi.node.base.AccountID toAccountId(final EntityId ent
                 .build();
     }
 
+    public static com.hedera.hapi.node.base.AccountID toAccountId(final Long shard, final Long realm, final Long num) {
```
[the remark this answers is not in the document]
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/services/utils/EntityIdUtils.java:189]: Comment addressed

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/CommonEntityAccessor.java [hunk 17]
```diff
@@ -40,6 +42,21 @@ public class CommonEntityAccessor {
...
+
+    public AccountID getAccountWithCanonicalAddress(EntityId entityId, final Optional<Long> timestamp) {
+        final var entity = timestamp
+                .map(t -> entityRepository
+                        .findActiveByIdAndTimestamp(entityId.getId(), t)
+                        .orElse(null))
+                .orElseGet(() -> entityRepository
+                        .findByIdAndDeletedIsFalse(entityId.getId())
+                        .orElse(null));
+        return toAccountId(entity);
```
[the remark this answers is not in the document]
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/CommonEntityAccessor.java]: Comment addressed

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+                .paused(token.getPauseStatus().equals(TokenPauseStatusEnum.PAUSED))
+                .supplyKey(Utils.parseKey(token.getSupplyKey()))
+                .supplyType(TokenSupplyType.valueOf(token.getSupplyType().name()))
+                .symbol(token.getSymbol())
+                .tokenId(toTokenId(entity.getId()))
+                .tokenType(TokenType.valueOf(token.getType().name()))
+                .totalSupply(getTotalSupply(token, timestamp))
+                .treasuryAccountId(getTreasury(token.getTreasuryAccountId(), timestamp))
+                .customFees(getCustomFees(token.getTokenId(), timestamp))
+                .pauseKey(Utils.parseKey(token.getPauseKey()))
```
[the remark this answers is not in the document]
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java]: Comment addressed

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+    }
+
+    private Supplier<AccountID> getAutoRenewAccount(Long autoRenewAccountId, final Optional<Long> timestamp) {
+        if (autoRenewAccountId == null) {
+            return null;
+        }
+        return Suppliers.memoize(() -> timestamp
+                .map(t -> entityRepository.findActiveByIdAndTimestamp(autoRenewAccountId, t))
+                .orElseGet(() -> entityRepository.findByIdAndDeletedIsFalse(autoRenewAccountId))
+                .map(EntityIdUtils::toAccountId)
```
[the remark this answers is not in the document]
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:158]: The main reason is that we are currently just mirroring the TokenDatabaseAccessor logic which is also querying for the entities for autorenewAccount and treasury.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+                .feeScheduleKey(Utils.parseKey(token.getFeeScheduleKey()))
+                .freezeKey(Utils.parseKey(token.getFreezeKey()))
+                .kycKey(Utils.parseKey(token.getKycKey()))
+                .maxSupply(token.getMaxSupply())
+                .memo(entity.getMemo())
+                .metadata(Bytes.wrap(token.getMetadata()))
+                .metadataKey(Utils.parseKey(token.getMetadataKey()))
+                .name(token.getName())
+                .pauseKey(Utils.parseKey(token.getPauseKey()))
+                .paused(token.getPauseStatus().equals(TokenPauseStatusEnum.PAUSED))
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:127]: `paused` should be before `pauseKey`

Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:131]: Comment addressed

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (IvanKavaldzhiev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:168]: We can see if we are going to need the whole entity models for other ReadableKVStates. If not, we can really simplify the queries as you suggested.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java [hunk 21]
```diff
@@ -0,0 +1,262 @@
...
+    private Supplier<Long> getTotalSupply(
+            final com.hedera.mirror.common.domain.token.Token token, final Optional<Long> timestamp) {
+        return Suppliers.memoize(() -> timestamp
+                .map(t -> getTotalSupplyHistorical(token.getTokenId(), t))
+                .or(() -> Optional.ofNullable(token.getTotalSupply()))
+                .orElse(0L));
+    }
+
+    private Long getTotalSupplyHistorical(long tokenId, long timestamp) {
+        return tokenRepository.findFungibleTotalSupplyByTokenIdAndTimestamp(tokenId, timestamp);
```
[the remark this answers is not in the document]
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java]: This was copied from initial PR. Maybe the initial idea was that NFT kv state will handle it but there is no such logic there. I've equalized the logic as in the tokenDatabaseAccessor.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:168]: We should create a ticket to capture these various future optimizations. There have been similar [comments](https://github.com/hashgraph/hedera-mirror-node/pull/9527#discussion_r1795085661) in other PRs about things we'd like to check later. Let's not forget these items.

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java [hunk 22]
```diff
@@ -0,0 +1,276 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import org.springframework.util.CollectionUtils;
+
+public class TokenReadableKVState extends ReadableKVStateBase<TokenID, Token> {
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:58]: Needs to have `@Named`

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java (hunk 22 above)
Comment (kselveliev) [reply hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/TokenReadableKVState.java:60]: Comment addressed

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Review by IvanKavaldzhiev — APPROVED** [review]
LGTM

**Comment by kselveliev** [comment]
Closing in favor of https://github.com/hashgraph/hedera-mirror-node/pull/9572
